datafusion/src/optimizer/single_distinct_to_groupby.rs

```diff
@@ -0,0 +1,261 @@
+///          F1(DISTINCT s0, s1, ...),
+///          F2(DISTINCT s0, s1, ...),
+///       - X
+///
+///   into
+///
+///   - Aggregation
+///            GROUP BY (k)
+///            F1(x)
+///            F2(x)
+///       - Aggregation
+///               GROUP BY (k, s0, s1, ...)
+///            - X
+///   </pre>
+///   <p>
+pub struct SingleDistinctToGroupBy {}
+
+impl SingleDistinctToGroupBy {
+    #[allow(missing_docs)]
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+fn optimize(plan: &LogicalPlan, execution_props: &ExecutionProps) -> Result<LogicalPlan> {
+    match plan {
+        LogicalPlan::Aggregate {
+            input,
+            aggr_expr,
+            schema: _,
+            group_expr,
+        } => {
+            match is_single_agg(plan) {
+                true => {
+                    let mut all_group_args: Vec<Expr> = Vec::new();
+                    all_group_args.append(&mut group_expr.clone());
+                    // remove distinct and collection args
+                    let mut new_aggr_expr = aggr_expr
+                        .iter()
+                        .map(|aggfunc| match aggfunc {
+                            Expr::AggregateFunction { fun, args, .. } => {
+                                all_group_args.append(&mut args.clone());
+                                Expr::AggregateFunction {
+                                    fun: fun.clone(),
+                                    args: args.clone(),
+                                    distinct: false,
+                                }
+                            }
+                            _ => aggfunc.clone(),
+                        })
+                        .collect::<Vec<_>>();
+
+                    let all_field = all_group_args
+                        .iter()
+                        .map(|expr| expr.to_field(input.schema()).unwrap())
+                        .collect::<Vec<_>>();
+
+                    let grouped_schema = Arc::new(DFSchema::new(all_field).unwrap());
+                    let new_aggregate = LogicalPlan::Aggregate {
+                        input: input.clone(),
+                        group_expr: all_group_args,
+                        aggr_expr: Vec::new(),
+                        schema: grouped_schema,
+                    };
+                    let mut expres = group_expr.clone();
+                    expres.append(&mut new_aggr_expr);
+                    utils::from_plan(plan, &expres, &[new_aggregate])
+                }
+                false => {
+                    let expr = plan.expressions();
+                    // apply the optimization to all inputs of the plan
+                    let inputs = plan.inputs();
+
+                    let new_inputs = inputs
+                        .iter()
+                        .map(|plan| optimize(plan, execution_props))
+                        .collect::<Result<Vec<_>>>()?;
+
+                    utils::from_plan(plan, &expr, &new_inputs)
```

Comment:
This part is redundant from the 113~119 lines, so if we can eliminate duplicate code.

Reply:
Thanks i fixed it